internal/component/redis/redis.go

```diff
@@ -37,110 +79,71 @@ func ParseClientFromProperties(properties map[string]string, defaultSettings *Se
-		MasterName:         s.SentinelMasterName,
-		SentinelAddrs:      []string{s.Host},
-		Password:           s.Password,
-		Username:           s.Username,
-		MaxRetries:         s.RedisMaxRetries,
-		MaxRetryBackoff:    time.Duration(s.RedisMaxRetryInterval),
-		MinRetryBackoff:    time.Duration(s.RedisMinRetryInterval),
-		DialTimeout:        time.Duration(s.DialTimeout),
-		ReadTimeout:        time.Duration(s.ReadTimeout),
-		WriteTimeout:       time.Duration(s.WriteTimeout),
-		PoolSize:           s.PoolSize,
-		MaxConnAge:         time.Duration(s.MaxConnAge),
-		MinIdleConns:       s.MinIdleConns,
-		PoolTimeout:        time.Duration(s.PoolTimeout),
-		IdleCheckFrequency: time.Duration(s.IdleCheckFrequency),
-		IdleTimeout:        time.Duration(s.IdleTimeout),
+	version, versionErr := GetServerVersion(c)
+	c.Close() // close the client to avoid leaking connections
+
+	useNewClient := false
+	if versionErr != nil {
+		// we couldn't query the server version, so we will assume the v8 client is not supported
+		useNewClient = true
+	} else if semver.Compare(version, "7.0.0") > -1 {
+		// if the server version is >= 7, we will use the v9 client
+		useNewClient = true
 	}
-
-	/* #nosec */
-	if s.EnableTLS {
-		opts.TLSConfig = &tls.Config{
-			InsecureSkipVerify: s.EnableTLS,
+	if useNewClient {
+		if settings.Failover {
+			return newV9FailoverClient(settings), settings, nil
 		}
+		return newV9Client(settings), settings, nil
+	} else {
+		if settings.Failover {
+			return newV8FailoverClient(settings), settings, nil
+		}
+		return newV8Client(settings), settings, nil
 	}
-
-	if s.RedisType == ClusterType {
-		opts.SentinelAddrs = strings.Split(s.Host, ",")
-
-		return redis.NewFailoverClusterClient(opts)
-	}
-
-	return redis.NewFailoverClient(opts)
 }
 
-func newClient(s *Settings) redis.UniversalClient {
-	if s == nil {
-		return nil
+func ClientHasJSONSupport(c RedisClient) bool {
+	if clientHasJSONSupport != nil {
+		return *clientHasJSONSupport
 	}
-	if s.RedisType == ClusterType {
-		options := &redis.ClusterOptions{
-			Addrs:              strings.Split(s.Host, ","),
-			Password:           s.Password,
-			Username:           s.Username,
-			MaxRetries:         s.RedisMaxRetries,
-			MaxRetryBackoff:    time.Duration(s.RedisMaxRetryInterval),
-			MinRetryBackoff:    time.Duration(s.RedisMinRetryInterval),
-			DialTimeout:        time.Duration(s.DialTimeout),
-			ReadTimeout:        time.Duration(s.ReadTimeout),
-			WriteTimeout:       time.Duration(s.WriteTimeout),
-			PoolSize:           s.PoolSize,
-			MaxConnAge:         time.Duration(s.MaxConnAge),
-			MinIdleConns:       s.MinIdleConns,
-			PoolTimeout:        time.Duration(s.PoolTimeout),
-			IdleCheckFrequency: time.Duration(s.IdleCheckFrequency),
-			IdleTimeout:        time.Duration(s.IdleTimeout),
-		}
-		/* #nosec */
-		if s.EnableTLS {
-			options.TLSConfig = &tls.Config{
-				InsecureSkipVerify: s.EnableTLS,
-			}
-		}
+	err := c.DoErr(context.Background(), "JSON.GET")
```

Comment:
SHould this have a timeout?

Reply:
I refactored the existing code. Right now my goal wasn't to add new functionality such as adding timeouts where there were none. We can do that in another pass if desired though.